[CALCITE-6162] Add rule(s) to remove joins with constant single tuple…

HanumathRao opened the pull request:
… relations

This PR introduces four rules for simplifying joins with a single value row on either side of the input. Please find the information about this optimizations in the related [JIRA](https://issues.apache.org/jira/browse/CALCITE-6162).
**Testing**: Apart from the new tests added in RelOptRulesTest, these rules are enabled as part of the abstract rules and so they get triggered during the core quidem test suite.

mihaibudiu commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -4177,6 +4178,109 @@ RelOptFixture checkDynamicFunctions(boolean treatDynamicCallsAsConstant) {
         .check();
   }
 
+  @Test void testJoinWithConstantRowOnRight() {
```
How about a test where some of the `VALUE`s are NULL?

HanumathRao replied:
These tests are designed to validate the generated plans for a specified query. I've incorporated output oriented tests into the join.iq file, and couple of them generate null values in the context of a left outer join. If you think any test cases are missing in join.iq, please let me know, and I'll be sure to add them. Thank you.

mihaibudiu replied:
equi-joining with a tuple that has a null value for one of the ON fields should produce an empty set. Is there a test about that?

HanumathRao replied:
I have added two test cases with null values as input. Please take a look and let me know. Thank you.

mihaibudiu replied:
I meant a test like this one: `select e.empno, e.ename, c.* from emp e cross join (select 5, NULL) c`
This test seems to produce an incorrect plan because you just substitute the NULL for the corresponding column:

```
LogicalProject(EMPNO=[$0], ENAME=[$1], EXPR$0=[$9], EXPR$1=[$10])
  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8], $f9=[5], $f10=[null:NULL])
    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```

I believe that the result should be the empty set.

HanumathRao replied:
@mihaibudiu  Thanks for the review, the above query doesn't result in an empty subset as it is a cross join. I have added the test case and made sure that the results are same with and without the optimization (by removing the rules  and executing the test case).

JiajunBernoulli commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,361 @@
+  /** Configuration for a rule that simplifies join node with a project on a constant row
+   *  on its left input. */
+  @Value.Immutable
+  interface JoinLeftSingleValueRuleWithExprConfig extends PruneSingleValueRule.Config {
```
`JoinLeftSingleValueRuleWithExprConfig` and `JoinRightSingleValueRuleWithExprConfig` has many same code.

Can we extract them?

macroguo-ghy replied:
Agree.

HanumathRao replied:
Thanks @JiajunBernoulli  and @macroguo-ghy  for the suggestion. I moved the redundant code to onMatch base class function.

macroguo-ghy commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,361 @@
+      return new SingleValueOptimizationRules.PruneSingleValueRule(this) {
+        @Override public void onMatch(RelOptRuleCall call) {
+          Predicate<Join> predicate = elibilityPredicate(false);
+          Join join = call.rel(0);
```
Use `final` modifier

HanumathRao commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,361 @@
+        Join join, List<RexNode> rexNodes, RelNode otherNode,
+        Predicate<Join> nonTransformable, boolean isValuesLeftChild,
+        BiFunction<RexNode, List<RexNode>, List<RexNode>> litTransformer) {
+      this.relNode = otherNode;
```
@macroguo-ghy  sorry didn't quite get this. If the parameters don't match then it will result in compilation failure right?

mihaibudiu replied:
More importantly, I think you should add JavaDoc for the parameters describing what they mean.

HanumathRao commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,361 @@
+      }
+    }
+
+    static final Predicate<Join> elibilityPredicate(boolean isLeft) {
```
I thought about it but left it for now as I plan to add explore for semi join and anti join in a different JIRA.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,329 @@
+ * Collection of rules which simplify sections of query plan which are known to
+ * produce single row.
+ *
+ * <p>Conventionally, the way to represent a single row relational expression is
```
Values can represent only constant relations. So this comment is not entirely accurate.
I think a more precise JavaDoc comment would be that these rules handle joins where one of the relations is a constant relation (VALUES) with a single row. The existing comment implies that you are handling all cases that can generate a single row.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,329 @@
+  public static final RelOptRule JOIN_LEFT_PROJECT_INSTANCE =
+      SingleValueOptimizationRules.JoinLeftSingleValueRuleWithExprConfig.DEFAULT.toRule();
+
+  public static final RelOptRule JOIN_RIGHT_PROJECT_INSTANCE =
```
Can you add a Javadoc comment on these describing how they differ from each other?

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,329 @@
+      SingleValueOptimizationRules.JoinRightSingleValueRuleWithExprConfig.DEFAULT.toRule();
+
+  /**
+   * Transformer class to transform a single value nodes on either side of the join.
```
you should probably capitalize VALUES, since it's a SQL construct.
Does "single" here mean "with a single tuple"?

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,329 @@
+      }
+    }
+
+    static Predicate<Join> elibilityPredicate(boolean isLeft) {
```
the word is "eligibility".

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,329 @@
+      this.literals = rexNodes;
+    }
+
+    public @Nullable RelNode transform(RelBuilder relBuilder) {
```
I would give here a more elaborate comment. What happens is that a JOIN is converted into a FILTER where the filter predicate uses the join condition. I would in fact give a small example of a plan before and after.

Another very subtle point is that this works even if the VALUES contain NULL entries. This is because this generates code with comparisons to NULL, and comparing a value to NULL using `=` always returns `false`, so the result is an empty collection, as expected.

mihaibudiu replied:
Can you please add a comment in the code containing the second paragraph in my comment? I had to run in the debugger to understand why this happens.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,329 @@
+    private final ImmutableBitSet bitSet;
+    private final List<RexNode> fieldValues;
+    private final int offset;
+    ReplaceExprWithConstValue(ImmutableBitSet bitSet, List<RexNode> values, int offset) {
```
documenting parameters would help read the code.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,329 @@
+      }
+    }
+
+    protected void onMatch(RelOptRuleCall call, Values values,
```
please document these parameters as well. It's not clear what "project" for example is.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java:
```diff
@@ -0,0 +1,329 @@
+    }
+  }
+
+  /** Configuration for a rule that simplifies join node with a project on a constant row
```
what is a project on a constant row? Is it a projection applied to a VALUES node?
Isn't that case handled by a separate optimization step?

mihaibudiu replied:
This should be a comment in the code.

mihaibudiu replied:
please add this as a comment in the code too.

HanumathRao commented:
@mihaibudiu  Thanks for the suggestion. I have added the javadoc comments wherever I think are needed for clarity. Regarding the comments for the rules, I have already added comments for the rule configs so I think it might be redundant to add them again for the rules. Please take a look and let me know if there is anything not clear or missing.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValuesOptimizationRules.java:
```diff
@@ -0,0 +1,374 @@
+            .map(shuttle::apply)
+            .collect(Collectors.toList());
+      } else {
+        rexNodes = new ArrayList<>(values.tuples.get(0));
```
Maybe you can factor out this as a helper function.
You previously used getTuples(), here you just used tuples.
This could be called getSingle(Values) and could also assert that values has a single tuple.

HanumathRao replied:
I changed the code to use tuples in both the cases. I haven't created a helper function here because it might be an overkill.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValuesOptimizationRules.java:
```diff
@@ -0,0 +1,374 @@
+        rexNodes = new ArrayList<>(values.tuples.get(0));
+      }
+      RelBuilder relBuilder = call.builder();
+      RelNode transformed =
```
I find that it's more convenient for debugging to have one statement per line.
So I would make a separate statement with the new and one with 'transform'.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValuesOptimizationRules.java:
```diff
@@ -0,0 +1,374 @@
+
+    static Predicate<Join> eligibilityPredicate(boolean isLeft) {
+      if (isLeft) {
+        return jn -> jn.getJoinType() == JoinRelType.LEFT
```
This may be clearer if you factor out the common case outside the 'if'.
One other thing that is confusing is that this is actually the "nonEligibilityPredicate".
Perhaps it would be clearer if you actually complement all definitions and uses of this predicate.
Then you can call the previous field "canTransform".

mihaibudiu reviewed (commented):
The code looks pretty good, my comments are about readability mostly.
I am not familiar with this part of the codebase; perhaps a more experienced reviewer would understand faster.
But in general I really prefer the code to be as simple and clear as possible. It helps everyone: reviewers, people who debug it, people who edit it in the future. Some of my suggestions may be not strictly necessary, like reverting the negated variables "cannotTransform", but I think they do really help.
Some of the comments you made in the PR should probably be in the code as well; the PR will be lost, but the code stays.

HanumathRao commented on core/src/test/java/org/apache/calcite/test/JdbcTest.java:
```diff
@@ -2973,27 +2973,29 @@ void testInnerJoinValues(String format) {
+          + "expr#0..1 = {inputs}\\n"
+          + "expr#2 = 10\\n"
+          + "expr#3 = 'SameName'\\n"
+          + "expr#4 = CAST($t1):I\\nNTEGER NOT NULL\\n"
```
I believe it is due to formatting (when dot format is used). There are other cases in the same file where \\n was in middle of classname or fieldname 
e.g. ""EnumerableTableScan\ntable" and "$t1\n0"

HanumathRao commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValuesOptimizationRules.java:
```diff
@@ -0,0 +1,374 @@
+            rexBuilder.makeCall(SqlStdOperatorTable.CASE, condition,
+                lit, rexBuilder.makeNullLiteral(lit.getType()))).collect(Collectors.toList());
+      default:
+        return (condition, rexLiterals) -> new ArrayList<>(rexLiterals);
```
No need to copy.

HanumathRao commented:
Not sure why the tests are showing failed in the PR but all of them are successful on the branch commit.
https://github.com/HanumathRao/calcite/commits/CALCITE-6162/, build run https://github.com/HanumathRao/calcite/actions/runs/7521737012/job/20472927213

HanumathRao commented:
> The code looks pretty good, my comments are about readability mostly. I am not familiar with this part of the codebase; perhaps a more experienced reviewer would understand faster. But in general I really prefer the code to be as simple and clear as possible. It helps everyone: reviewers, people who debug it, people who edit it in the future. Some of my suggestions may be not strictly necessary, like reverting the negated variables "cannotTransform", but I think they do really help. Some of the comments you made in the PR should probably be in the code as well; the PR will be lost, but the code stays.

Thank you for your review comments. I have addressed the comments and provided inline replies where necessary.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValuesOptimizationRules.java:
```diff
@@ -95,24 +95,26 @@ private static class SingleValuesRelTransformer {
     protected SingleValuesRelTransformer(
         Join join, List<RexNode> rexNodes, RelNode otherNode,
-        Predicate<Join> nonTransformable, boolean isValuesLeftChild,
+        Predicate<Join> tranfsormable, boolean isValuesLeftChild,
```
parameter name is mistyped "sf"

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SingleValuesOptimizationRules.java:
```diff
@@ -245,24 +256,29 @@ protected void onMatch(RelOptRuleCall call, Values values,
           new SingleValuesRelTransformer(join, rexNodes, other,
-              predicate, isLeft, getRexTransformer(relBuilder.getRexBuilder(),
-              join.getJoinType())).transform(relBuilder);
+              transformableCheck, isOnLefSide, transformer).transform(relBuilder);
```
At least with IntelliJ I have trouble stepping in "transform" here.
So I would even allocate a temp variable for the transformer and invoke `transform` in a separate statement.

mihaibudiu commented:
I have approved, but I still made a few small suggestions.
It looks like the CI is unhappy.

HanumathRao commented:
> I have approved, but I still made a few small suggestions. It looks like the CI is unhappy.

Thanks. I am not sure why only PR runs are showing this issue. The tests executed on the commit and also same commit on another branch are always successful. I am guessing that once I squash all the commits and rebase this problem might go away.

mihaibudiu commented:
I have no further comments. Please squash the commits so we can merge this PR. Thank you for the improvements!

HanumathRao commented:
> I have no further comments. Please squash the commits so we can merge this PR. Thank you for the improvements!

Done. Thanks for review.
